Updates to the Umbraco Forms developer documentation (#4643)
umbraco/UmbracoDocs: Add-ons/UmbracoForms/Developer/Rendering-Scripts/index.md

# Rendering Forms Scripts

> Script delivery is done automatically using [Themes](/documentation/Add-ons/UmbracoForms/Developer/Themes/) without the need to include any additional code into your own views. The 'default' theme (bundled with Umbraco Forms) comes with the file `Scripts.cshtml` containing all the relevant logic & data required by the Umbraco Forms JavaScript. For reference, the file looks like this *(as per Umbraco Forms 10.2)*:

Reviewer: This is great to add, but I think we should remove the "For reference... " sentence, and the code block.  It'll only get out of date here.  I'd rather direct people to where they can download the theme files.

     data-serialized-field-conditions="@JsonConvert.SerializeObject(Model.FieldConditions, FormsJsonSerializerSettings.Default)"
     data-serialized-fields-not-displayed="@JsonConvert.SerializeObject(Model.GetFieldsNotDisplayed(), FormsJsonSerializerSettings.Default)"
     data-trigger-conditions-check-on="@Model.TriggerConditionsCheckOn"
     data-form-element-html-id-prefix="@Model.FormElementHtmlIdPrefix"></div>

@* Only render out scripts on the page if the form has not been submitted yet *@
@if (Model.SubmitHandled == false)
{
    @*
        If the current page of the form has any Partial view files attached to render
        Likely used by events and third parties adding tracking or other 3rd party functionality to a form
    *@
    if (Model.CurrentPage.PartialViewFiles.Any())
    {
        foreach (var partial in Model.CurrentPage.PartialViewFiles)
        {
            @await Html.PartialAsync(partial.Value)
        }
    }

    @* Render references to javascript files needed by fields on the current page*@
    @Html.RenderFormsScripts(Url, Model, Model.JavaScriptTagAttributes)
    @Html.RenderFormsStylesheets(Url, Model)
}

```

:::note

> If you are upgrading from an older version of Umbraco Forms and using a variant of `Scripts.cshtml` within your own project you may be missing some required and fundamental parts of this file. As an example, the `<div class="umbraco-forms-form-config" />` element introduced the `data-form-element-html-id-prefix` attribute in Umbraco Forms 10.2. Failing to update this within your own version of this file may lead to a loss of functionality and/or JavaScript errors.

Reviewer 2: ```suggestion
When upgrading from an older version of Umbraco Forms and using a variant of `Scripts.cshtml`, you may be missing required parts of this file. As an example, the `<div class="umbraco-forms-form-config" />` element introduced the `data-form-element-html-id-prefix` attribute in Umbraco Forms 10.2. Failing to update this within your own version of this file may lead to a loss of functionality and/or JavaScript errors.
```

> If you are upgrading from an older version of Umbraco Forms and using a variant of `Scripts.cshtml` within your own project you may be missing some required and fundamental parts of this file. As an example, the `<div class="umbraco-forms-form-config" />` element introduced the `data-form-element-html-id-prefix` attribute in Umbraco Forms 10.2. Failing to update this within your own version of this file may lead to a loss of functionality and/or JavaScript errors.

Reviewer: This is fair point, but very specific about a particular version and issue - and really the point is general.  If you're using a custom theme, you will want to review when the default one is updated to make sure you apply to your custom theme what changes may be necessary

Rather than adding this, probably something to that effect can be usefully added [here](https://our.umbraco.com/documentation/Add-ons/UmbracoForms/Developer/Themes/#creating-a-theme).

`ExcludeScripts = "1"` prevents the associated scripts from being rendered. Any other value, an empty value, or if the parameter is excluded, will render the scripts on the Form.
:::

> The `Scripts.cshtml` file (from the current or default theme) will not be initiated and won't be included as part of the current form. You may however use parts (or all) of the code from the snippet of `Scripts.cshtml` within your own template or partial views.

Reviewer 2: ```suggestion
The `Scripts.cshtml` file (from the current theme) will not be initiated and therefore not included as part of the current form. You may however use parts (or all) of the code from the snippet of `Scripts.cshtml` within your own template or partial views.
```

> In many cases, you might prefer rendering your scripts at the bottom of the page, e.g. before the closing `</body>` tag, as this generally improves site performance.

Reviewer 2: ```suggestion
In many cases, you might prefer rendering your scripts at the bottom of the page, for example before the closing `</body>` tag. This will generally improve site performance.
```
